File: 1143.js
/**
 * @param {string} text1
 * @param {string} text2
 * @return {number}
 */
var longestCommonSubsequence = function(text1, text2) {

    let dp = [];
    for (let i = 0; i < text1.length; i++) {
        let row = [];
        for (let j = 0; j < text2.length; j++) {
            row.push(null);
        }
        dp.push(row);
    }
    for (let i = text1.length-1; i>=0; i--) {
        for (let j = text2.length-1; j>=0; j--) {
            let result = null;
            if (i === text1.length-1 || j === text2.length-1) {
                result = (text1[i] === text2[j] || (dp[i] && dp[i][j+1] === 1) || dp[i+1] && dp[i+1][j] === 1) ? 1 : 0;
            } else {
                if (text1[i] === text2[j]) {
                    result = dp[i+1][j+1] + 1;
                } else {
                    result = Math.max(dp[i+1][j], dp[i][j+1]);
                }
            }
            dp[i][j] = result;
        }
    }
    return dp[0][0];
};

File: 1143.py
class Solution:
    def LCS_helper(self, text1, text2, start1, start2, dp) -> int:
        if (start1 == len(text1) or start2 == len(text2)):
            return 0
        if (dp[start1][start2] is not None):
            return dp[start1][start2]
        sol = None
        if text1[start1] == text2[start2]:
            sol = 1 + self.LCS_helper(text1, text2, start1+1, start2+1, dp) 
        else:
            sol = max(
                self.LCS_helper(text1, text2, start1, start2+1, dp),
                self.LCS_helper(text1, text2, start1+1, start2, dp))
        dp[start1][start2] = sol
        return sol

    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = [[None]*len(text2) for i in range(len(text1))]
        return self.LCS_helper(text1, text2, 0, 0, dp)